### PR TITLE
Feature/app store webhook

### PR DESCRIPTION
**Pull request by sc-ecolasurdo**
## Summary
Add endpoints to cover App Store login flows.
Add WebHook action and methods to update FBE data using WebHook events.

## Note
Test's last step depends on Connect Bridge to be live.

## UI Changes
N/A

### Setup:
- Install Facebook for WooCommerce from this PR.

### Steps:
1. Browse this page after replacing {PAGE_ID} with your Facebook Page ID:
https://connectbridge.ngrok.io/app-store-login/facebook/?redirect_uri=https%3A%2F%2Fwww.facebook.com%2Flocal%2Fdev%2Ftransaction_tool_selector%2Fredirect%3Fapp_id%3D474166926521348%26app_name%3DWooCommerce%2BIntegration%26page_id%3D{PAGE_ID}
2. Connect Bridge will ask for your Store URL. Provide your Store URL and press connect.
3. Grant access to your site.
4. Browse Admin > Marketing > Facebook to see if you are connected.

### CODE REVIEW

File: includes/Handlers/Connection.php [hunk 1]
```diff
@@ -1032,4 +1083,260 @@ public function get_plugin() {
...
+			wp_safe_redirect( site_url() );
+			exit;
+		}
+
+		if ( empty( $_REQUEST['success'] ) ) { //phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+			$url_params = [
+				'store_url'    => '',
+				'redirect_uri' => rawurlencode( $redirect_uri ),
+				'errors'       => [ 'You need to grant access to Facebook.' ],
```
Comment (jconroy) [inline includes/Handlers/Connection.php:1308]: Should this actually be "WooCommerce"

**Review by jconroy — COMMENTED** [review]
Nice, thanks @sc-ecolasurdo - I managed to get through the process with a local install of  the connect server up to being redirected back to Facebook - at which point I get a white screen (on the facebook side) - but all the values/keys saved within the connect server database

File: includes/Handlers/Connection.php [hunk 2]
```diff
@@ -1032,4 +1083,260 @@ public function get_plugin() {
...
+	 * @return array
+	 */
+	public function change_permissions( $permissions ) {
+
+		if ( empty( $_REQUEST['app_name'] ) || 'WooCommerce Integration' !== $_REQUEST['app_name'] ) { //phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			return $permissions;
+		}
+
+		return array(
+			__( 'Update FBE Installation data', 'facebook-for-wordpress' ),
```
[the remark this answers is not in the document]
Comment (sc-ecolasurdo) [reply includes/Handlers/Connection.php]: Yes, I agree, will send the change. Will be more clear for the merchant.

File: includes/Handlers/Connection.php [hunk 3]
```diff
@@ -1032,4 +1083,260 @@ public function get_plugin() {
...
+	 * Endpoint permissions
+	 * Woo Connect Bridge is sending the WebHook request using generated key.
+	 *
+	 * @return boolean
+	 */
+	public function extras_permission_callback() {
+
+		add_filter( 'woocommerce_rest_is_request_to_rest_api', '__return_true' );
+
+		$user = apply_filters( 'determine_current_user', null );
```
[the remark this answers is not in the document]
Comment (sc-ecolasurdo) [reply includes/Handlers/Connection.php:1266]: The intention is to authenticate the user using Basic auth.
This way we trigger _authenticate_ method from WC_REST_Authentication.
We need to use the _ woocommerce_rest_is_request_to_rest_api_ filter so is_request_to_rest_api() returns true.

**Comment by jconroy** [comment]
Some more feedback from internal

* function `parse_body` doesn't look like it's needed, [WP_REST_Request](https://developer.wordpress.org/reference/classes/wp_rest_request/) has a ton of functions to get the data from the request.
* `webhook_callback` parses the body twice, we can get it directly from the `WP_REST_Request` and check if it's a valid object before passing it to the hook (no need for reading from php://input)
* Rest API endpoints return empty on success and fail is that intended to never provide feedback?
* Permissions for both REST API requests seem a little odd, I'm not really following what they are trying to do there but that doesn't look right at all. If they want to allow those specific requests by checking a different auth header then they should be hooking into `determine_current_user` to validate / check the header.
* What's the purpose of `change_permissions`, that looks like it's hooking into the WC Auth endpoint, why is that even being used?
* Hardcoded https://connect.woocommerce.com/app-store-login/facebook, can we use a filter similar to wc_facebook_connection_proxy_url
* The function `extras_callback` shouldn't use `wp_send_json`. It should either return a `WP_REST_Response` containing the data or just the JSON data directly which will be wrapped in a response.

Also

> I guess for point 4 the site's CK/CS is being saved in connect.woocommerce.com so it must be using that to talk to the site. Which means they are trying to get the API request with a prefix of facebook/v1 to authenticate as a WC API request. I think they should use `wc-facebook/v1` instead which is allowed here: https://github.com/woocommerce/woocommerce/blob/4.9.2/includes/class-wc-rest-authentication.php#L64-L65
Once that is set they can do normal permission checks based on what the key was granted, maybe:  `current_user_can( 'manage_woocommerce' )`

File: includes/Handlers/Connection.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jconroy) [reply includes/Handlers/Connection.php:1266]: Thanks, [similar query/feedback here](https://github.com/facebookincubator/facebook-for-woocommerce/pull/1764#issuecomment-776180914)

File: includes/Handlers/Connection.php [hunk 4]
```diff
@@ -736,6 +762,26 @@ public function get_proxy_url() {
...
+	 * @since 2.2.1-dev.1
+	 *
+	 * @return string URL
+	 */
+	public function get_app_store_login_url() {
+
+		/**
+		 * Filters App Store login URL.
+		 *
+		 * @since 2.0.0
```
Comment (jconroy) [inline includes/Handlers/Connection.php:777]: 2.2.1-dev.1

**Review by jconroy — COMMENTED** [review]
Thanks for the latest edits @sc-ecolasurdo just picked up one little thing. ~I'm going to take it for a test run in a little while~

Actually what the best way to try the flow out now?

**Pull request merged**